pagefaultgames/pokerogue
# [QoL] Run History Function

## What are the changes?
This adds a new option on the menu for 'Run History'.
It allows users to store and look at a maximum of 25 past runs. This functionality automatically logs a run based on the current SessionSaveData and stores it in cache/the network database. At the moment, the function mirrors the 'Load Save' UI and only shows the team, whether they were victorious or not, the mode, wave, and the final timestamp of the run. 

The server-side code is here: https://github.com/pagefaultgames/rogueserver/pull/33

## Why am I doing these changes?
This has been a feature requested by many users. In addition, this will allow the Balance and Development teams to do some analytics and have data on the gameplay experience. 

I also have a bad memory so I thought it would be helpful to be able to access past runs to see what I need to do better. 

## What did change?
There are many changes:
- New Menu Option
- New UI
- New 'RunHistoryData' item in cache
- Network communication with new endpoint / table in database
- New post-battle behavior

### Screenshots/Videos
**Features**
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/1e3e6852-ac57-4193-b26e-cc1bdd68fb6a)
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/3f3c7e23-65ab-4441-9c8e-9c03d8753188)


**Languages**
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/c5bc0a8c-82eb-469d-825a-416cc1382c1a)
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/5feeab05-7afc-4e41-b785-d8ebae8991c4)
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/b947f664-cfd5-4460-883c-a528f77e9f03)
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/3bc2716f-8940-40ba-9b70-6fd2ac5b736d)
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/a4da6c47-9e33-4e4f-a71e-563dd59447aa)
![image](https://github.com/pagefaultgames/pokerogue/assets/171087428/5c8da076-a2cd-4ab7-8524-3ad76e53651c)



## How to test the changes?
Q: How can a reviewer test your changes once they check out on your branch? 
A: Play normally and log runs - successful or losses. 
Q: Did you just make use of the `src/overrides.ts` file?
A: Yes - but just to kill my Pokemon faster so I could log defeats. 
Q: Did you introduce any automated tests?
A: No
Q: Does the reviewer need to do something special in order to test your change?
A: Yes - need to run RogueServer for network/database functionality

## Checklist
    This still does not pass the Linter, but I plan on cleaning the PR up once all changes are made. My code is really messy, so I feel hesitant to remove all the imports.
